src/values/ObjectValue.js

```diff
@@ -645,25 +645,79 @@ export default class ObjectValue extends ConcreteValue {
     }
 
     if (!(P instanceof AbstractValue)) return this.$Get(P, Receiver);
+
+    // A string coercion might have side-effects.
+    // TODO #1682: We assume that simple objects mean that they don't have a
+    // side-effectful valueOf and toString but that's not enforced.
+    if (P.mightNotBeString() && P.mightNotBeNumber() && !P.isSimpleObject()) {
+      if (this.$Realm.isInPureScope()) {
+        // If we're in pure scope, we can havoc the key and keep going.
+        // Coercion can only have effects on anything reachable from the key.
+        Havoc.value(this.$Realm, P);
+      } else {
+        let error = new CompilerDiagnostic(
+          "property key might not have a well behaved toString or be a symbol",
+          this.$Realm.currentLocation,
+          "PP0002",
+          "RecoverableError"
+        );
+        if (this.$Realm.handleError(error) !== "Recover") {
+          throw new FatalError();
+        }
+      }
+    }
+
     // We assume that simple objects have no getter/setter properties.
-    if (
-      this !== Receiver ||
-      !this.isSimpleObject() ||
-      (P.mightNotBeString() && P.mightNotBeNumber() && !P.isSimpleObject())
-    ) {
-      // if P is an abstract value that we don't know about, but we're in pure scope
-      // then if the object is simple, then we can safely continue without throwing
-      // the introspection error below, since converting P to a string is assumed to
-      // be well behaved in a pure scope
-      if (!(this.$Realm.isInPureScope() && this.isSimpleObject() && this === Receiver)) {
-        AbstractValue.reportIntrospectionError(P, "TODO: #1021");
-        throw new FatalError();
+    if (!this.isSimpleObject()) {
+      if (this.$Realm.isInPureScope()) {
+        // If we're in pure scope, we can havoc the object. Coercion
+        // can only have effects on anything reachable from this object.
+        // We assume that if the receiver is different than this object,
+        // then we only got here because there were no other keys with
+        // this name on other parts of the prototype chain.
+        // TODO #1675: A fix to 1675 needs to take this into account.
+        Havoc.value(this.$Realm, Receiver);
+        return AbstractValue.createTemporalFromBuildFunction(this.$Realm, Value, [Receiver, P], ([o, p]) =>
+          t.memberExpression(o, p, true)
+        );
+      } else {
+        let error = new CompilerDiagnostic(
+          "unknown property access might need to invoke a getter",
+          this.$Realm.currentLocation,
+          "PP0030",
+          "RecoverableError"
+        );
+        if (this.$Realm.handleError(error) !== "Recover") {
+          throw new FatalError();
+        }
       }
     }
+
+    if (P.mightNotBeString()) {
+      // If the property is not a string we need to coerce it.
```

Comment:
I also added this string coercion which fixes #1676.

It is also required when the key is not an object, which allows me to test the havocing behavior of these.